### PR TITLE
Conform RuntimeError to HTTPResponseConvertible

### PR DESCRIPTION
**Pull request by gayathrisairam**
### Motivation

https://github.com/apple/swift-openapi-generator/issues/609#issuecomment-2522503704

### Modifications

Confirm `RuntimeError` to `HTTPResponseConvertible` and provide granular status codes.

### Result

Response codes for bad user input will be 4xx (instead of 500)

### Test Plan

Unit tests.

### CODE REVIEW

File: Sources/OpenAPIRuntime/Errors/RuntimeError.swift [hunk 1]
```diff
@@ -141,3 +141,43 @@ internal enum RuntimeError: Error, CustomStringConvertible, LocalizedError, Pret
...
+                .invalidServerVariableValue:
+            .notFound
+        case .invalidExpectedContentType,
+                .missingCoderForCustomContentType,
+                .unexpectedContentTypeHeader:
+            .unsupportedMediaType
+        case .unexpectedAcceptHeader(_):
+            .notAcceptable
+        case .missingOrMalformedContentDispositionName:
+            .unprocessableContent
```
Comment (simonjbeaumont) [inline Sources/OpenAPIRuntime/Errors/RuntimeError.swift:159]: > The HTTP 422 Unprocessable Content [client error response](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses) status code indicates that the server understood the content type of the request content, and the syntax of the request content was correct, but it was unable to process the contained instructions.
— https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422


Sounds like this one might not be the correct response for missing or malformed. Should we return 401 instead here?

File: Sources/OpenAPIRuntime/Errors/RuntimeError.swift [hunk 2]
```diff
@@ -141,3 +141,43 @@ internal enum RuntimeError: Error, CustomStringConvertible, LocalizedError, Pret
...
+    public var httpStatus: HTTPTypes.HTTPResponse.Status {
+        switch self {
+        case .invalidServerURL,
+                .invalidServerVariableValue:
+            .notFound
+        case .invalidExpectedContentType,
+                .missingCoderForCustomContentType,
+                .unexpectedContentTypeHeader:
+            .unsupportedMediaType
+        case .unexpectedAcceptHeader(_):
```
Comment (simonjbeaumont) [inline Sources/OpenAPIRuntime/Errors/RuntimeError.swift:156]: Nit (probably will get picked up by the formatter): 
```suggestion
        case .unexpectedAcceptHeader:
```

File: Sources/OpenAPIRuntime/Errors/RuntimeError.swift [hunk 3]
```diff
@@ -141,3 +141,43 @@ internal enum RuntimeError: Error, CustomStringConvertible, LocalizedError, Pret
...
+                .invalidBase64String,
+                .invalidHeaderFieldName,
+                .malformedAcceptHeader,
+                .missingMultipartBoundaryContentTypeParameter,
+                .missingRequiredHeaderField,
+                .missingRequiredMultipartFormDataContentType,
+                .missingRequiredQueryParameter,
+                .missingRequiredPathParameter,
+                .missingRequiredRequestBody,
+                .pathUnset,
```
Comment (simonjbeaumont) [inline Sources/OpenAPIRuntime/Errors/RuntimeError.swift:171]: Not sure but opening discussion: should this be 404 instead?

File: Sources/OpenAPIRuntime/Errors/RuntimeError.swift [hunk 4]
```diff
@@ -141,3 +141,43 @@ internal enum RuntimeError: Error, CustomStringConvertible, LocalizedError, Pret
...
+/// HTTP Response status definition for ``RuntimeError``. 
+extension RuntimeError: HTTPResponseConvertible {
+    public var httpStatus: HTTPTypes.HTTPResponse.Status {
+        switch self {
+        case .invalidServerURL,
+                .invalidServerVariableValue:
+            .notFound
+        case .invalidExpectedContentType,
+                .missingCoderForCustomContentType,
+                .unexpectedContentTypeHeader:
```
Comment (simonjbeaumont) [inline Sources/OpenAPIRuntime/Errors/RuntimeError.swift:154]: I think that these two are more 422 IIUC?

File: Tests/OpenAPIRuntimeTests/Errors/Test_RuntimeError.swift [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+//===----------------------------------------------------------------------===//
+
+import HTTPTypes
+@_spi(Generated) @testable import OpenAPIRuntime
+import XCTest
+
+struct MockRuntimeErrorHandler: Sendable {
+    var failWithError: (any Error)? = nil
+    func greet(_ input: String) async throws -> String {
+        if failWithError != nil { throw failWithError! }
```
Comment (simonjbeaumont) [inline Tests/OpenAPIRuntimeTests/Errors/Test_RuntimeError.swift:22]:
```suggestion
        if let failWithError { throw failWithError }
```

File: Tests/OpenAPIRuntimeTests/Errors/Test_RuntimeError.swift [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            deserializer: { request, body, metadata in
+                let body = try XCTUnwrap(body)
+                return try await String(collecting: body, upTo: 10)
+            },
+            serializer: { output, _ in fatalError() }
+        )
+        XCTAssertEqual(response.0.status, .internalServerError)
+    }
+
+    func testRuntimeError_withUnderlyingErrorConfirming_returns500() async throws {
```
Comment (simonjbeaumont) [inline Tests/OpenAPIRuntimeTests/Errors/Test_RuntimeError.swift:52]: This test name sounds wrong?

**Review by simonjbeaumont — COMMENTED** [review]
Thanks @gayathrisairam! Added some comments.

File: Sources/OpenAPIRuntime/Errors/RuntimeError.swift (hunk 1 above)
Comment (czechboy0) [reply Sources/OpenAPIRuntime/Errors/RuntimeError.swift:159]: Or a generic 400 is also fine, if it's a malformed header

File: Sources/OpenAPIRuntime/Errors/RuntimeError.swift (hunk 4 above)
Comment (gayathrisairam) [reply Sources/OpenAPIRuntime/Errors/RuntimeError.swift:154]: Isn't unexpectedContentTypeHeader 415?

File: Sources/OpenAPIRuntime/Errors/RuntimeError.swift [hunk 7]
```diff
@@ -141,3 +141,26 @@ internal enum RuntimeError: Error, CustomStringConvertible, LocalizedError, Pret
...
+        case .invalidExpectedContentType, .unexpectedContentTypeHeader: .unsupportedMediaType
+        case .missingCoderForCustomContentType: .unprocessableContent
+        case .unexpectedAcceptHeader: .notAcceptable
+        case .failedToDecodeStringConvertibleValue, .invalidAcceptSubstring, .invalidBase64String,
+            .invalidHeaderFieldName, .malformedAcceptHeader, .missingMultipartBoundaryContentTypeParameter,
+            .missingOrMalformedContentDispositionName, .missingRequiredHeaderField,
+            .missingRequiredMultipartFormDataContentType, .missingRequiredQueryParameter, .missingRequiredPathParameter,
+            .missingRequiredRequestBody, .unsupportedParameterStyle:
+            .badRequest
+        case .pathUnset: .notFound
```
Comment (simonjbeaumont) [inline Sources/OpenAPIRuntime/Errors/RuntimeError.swift:160]: Nit: Can we group this with the others that are returning `.notFound`?

File: Sources/OpenAPIRuntime/Errors/RuntimeError.swift [hunk 8]
```diff
@@ -13,7 +13,7 @@
 //===----------------------------------------------------------------------===//
 import protocol Foundation.LocalizedError
 import struct Foundation.Data
-
+import HTTPTypes
```
Comment (simonjbeaumont) [inline Sources/OpenAPIRuntime/Errors/RuntimeError.swift:16]: Nit.

```suggestion
import HTTPTypes

```

File: Tests/OpenAPIRuntimeTests/Errors/Test_RuntimeError.swift [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        guard input == "hello" else { throw TestError() }
+        return "bye"
+    }
+
+    static let requestBody: HTTPBody = HTTPBody("hello")
+    static let responseBody: HTTPBody = HTTPBody("bye")
+}
+
+final class Test_RuntimeError: XCTestCase {
+    func testRuntimeError_withUnderlyingErrorNotConfirming_returns500() async throws {
```
Comment (simonjbeaumont) [inline Tests/OpenAPIRuntimeTests/Errors/Test_RuntimeError.swift:32]: Typo:
```suggestion
    func testRuntimeError_withUnderlyingErrorNotConforming_returns500() async throws {
```

File: Tests/OpenAPIRuntimeTests/Errors/Test_RuntimeError.swift [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+            deserializer: { request, body, metadata in
+                let body = try XCTUnwrap(body)
+                return try await String(collecting: body, upTo: 10)
+            },
+            serializer: { output, _ in fatalError() }
+        )
+        XCTAssertEqual(response.0.status, .internalServerError)
+    }
+
+    func testRuntimeError_withUnderlyingErrorConfirming_returnsCorrectStatusCode() async throws {
```
Comment (simonjbeaumont) [inline Tests/OpenAPIRuntimeTests/Errors/Test_RuntimeError.swift:52]: Typo
```suggestion
    func testRuntimeError_withUnderlyingErrorConforming_returnsCorrectStatusCode() async throws {
```

**Review by simonjbeaumont — COMMENTED** [review]
This LGTM now. There's just a couple of typos though, then it's good to land.

**Review by simonjbeaumont — APPROVED** [review]
Great work—thanks @gayathrisairam!

**Pull request merged**